cmd/repo-updater/repos/sources.go

```
@@ -670,6 +674,78 @@ func (s *PhabricatorSource) client(ctx context.Context) (*phabricator.Client, er
 	return s.cli, err
 }
 
+// A AWSCodeCommitSource yields repositories from a single AWS Code Commit
+// connection configured in Sourcegraph via the external services
+// configuration.
+type AWSCodeCommitSource struct {
+	svc  *ExternalService
+	conn *awsCodeCommitConnection
+}
+
+// NewAWSCodeCommitSource returns a new AWSCodeCommitSourceSource from the given external service.
+func NewAWSCodeCommitSource(svc *ExternalService, cf *httpcli.Factory) (*AWSCodeCommitSource, error) {
+	var c schema.AWSCodeCommitConnection
+	if err := jsonc.Unmarshal(svc.Config, &c); err != nil {
+		return nil, fmt.Errorf("external service id=%d config error: %s", svc.ID, err)
+	}
+	return newAWSCodeCommitSource(svc, &c, cf)
+}
+
+func newAWSCodeCommitSource(svc *ExternalService, c *schema.AWSCodeCommitConnection, cf *httpcli.Factory) (*AWSCodeCommitSource, error) {
+	conn, err := newAWSCodeCommitConnection(c, cf)
+	if err != nil {
+		return nil, err
+	}
+	return &AWSCodeCommitSource{svc: svc, conn: conn}, nil
+}
+
+// ListRepos returns all AWS Code Commit repositories accessible to all
+// connections configured in Sourcegraph via the external services
+// configuration.
+func (s AWSCodeCommitSource) ListRepos(ctx context.Context) (repos []*Repo, err error) {
+	rs, err := s.conn.listAllRepositories(ctx)
+	for _, r := range rs {
+		repos = append(repos, awscodecommitRepoToRepo(s.svc, r, s.conn))
+	}
+	return repos, err
+}
+
+// ExternalServices returns a singleton slice containing the external service.
+func (s AWSCodeCommitSource) ExternalServices() ExternalServices {
+	return ExternalServices{s.svc}
+}
+
+func awscodecommitRepoToRepo(
+	svc *ExternalService,
+	repo *awscodecommit.Repository,
+	conn *awsCodeCommitConnection,
+) *Repo {
+	urn := svc.URN()
+	cloneURL, err := conn.authenticatedRemoteURL(repo)
+	if err != nil {
+		log15.Warn("Error adding authentication to AWS Code Commit repository Git remote URL.", "error", err)
```

We'd likely want to return an error to the user here.

---

Done!

---

Just returning the error should be good enough? I would suspect the logs as just noisy.